framework/src/modules/chain/dpos/dpos.js

```diff
@@ -77,17 +79,99 @@ module.exports = class Dpos {
 
 	async onRoundFinish() {
 		const disposableDelegateList =
-			this.finalizedBlockRound - this.delegateListRoundOffset;
+			this.finalizedBlockRound -
+			this.delegateListRoundOffset -
+			this.delegateActiveRoundLimit;
 		await this.delegatesList.deleteDelegateListUntilRound(
 			disposableDelegateList,
 		);
 	}
 
+	async getActiveDelegateHeights(
+		numberOfRounds = 1,
+		{ tx, delegateListRoundOffset = this.delegateListRoundOffset },
+	) {
+		// TODO: Discuss reintroducing a caching mechanism to avoid fetching
+		// active delegate lists multiple times.
+		let delegateLists = await this.storage.entities.RoundDelegates.get(
+			{
+				sort: 'round:desc',
+				// limit: number or requested rounds + last 3 rounds
+				limit: numberOfRounds + this.delegateActiveRoundLimit,
+			},
+			tx,
+		);
+
+		/**
+		 * Reverting the RoundOffset value, so that
+		 * round no can actually reflect the real round number.
+		 */
+		delegateLists = delegateLists.map(list => {
+			// eslint-disable-next-line no-param-reassign
+			list.round += delegateListRoundOffset;
+			return list;
+		});
+
+		const delegates = {};
+		for (let i = 0; i < numberOfRounds; i += 1) {
+			const list = delegateLists[i];
+			const activeRound = list.round;
+
+			list.delegatePublicKeys.forEach(publicKey => {
```

Comment:
seems like purpose of this function is more `reduce` than `forEach`

Reply:
That's true but since it is inside a for loop I didn't want to reassign `delegates` variable and didn't use `reduce`.